src/mon/OSDMonitor.cc

```diff
@@ -11392,6 +11392,105 @@ bool OSDMonitor::prepare_command_impl(MonOpRequestRef op,
+
+    if (skip_metadata_lookup)
+      ss << "reweighting all OSDs using scaling factor "
+         << g_conf()->osd_crush_scaling_factor << ":";
+    else
+      ss << "reweighting all OSDs using scaling factor "
+         << g_conf()->osd_crush_scaling_factor
+         << " based on device size metadata:";
+
+    if (dry_run)
+      ss << "This is a dry run, no changes will be made. "
+         << "Use 'test-reweight-by-scaling-factor -o newcrush' "
+         << "then 'crushdiff compare -c newcrush' "
+         << "to check if this will cause data movement.";
+
+    CrushWrapper newcrush = _get_pending_crush();
+
+    for (int id = 0; id < newcrush.get_max_devices(); ++id) {
+        if (!newcrush.get_item_weight(id)) {
+          // weight is 0, skip
+          continue;
+        }
+        double new_weight = double(g_conf()->osd_crush_scaling_factor) *
+                                   newcrush.get_item_weightf(id));
+
+        if (!skip_metadata_lookup) {
+          map<string,string> meta;
+          err = load_metadata(id, meta, nullptr);
+          if (err < 0) {
+            ss << "failed to load metadata for osd." << id
+               << ": " << cpp_strerror(err);
+            goto reply_no_propose;
+          }
+
+          // osd statfs.total is bluestore_bdev_size + bluefs_db_size
+          uint64_t total_size = 0;
+          auto p1 = meta.find("bluestore_bdev_size");
+          if (p1 == meta.end()) {
+            ss << "osd." << id << " does not have bluestore_bdev_size metadata";
+            goto reply_no_propose;
+          }
+          total_size = std::stoull(p1->second);
+          auto p2 = meta.find("bluefs_db_size");
+          if (p2 == meta.end()) {
+            ss << "osd." << id << " does not have bluefs_db_size metadata";
+            goto reply_no_propose;
+          }
+          total_size += std::stoull(p2->second);
+
+          // use the same calculation as OSD::update_crush_location
+          new_weight = std::max(0.00001,
+                                double(g_conf()->osd_crush_scaling_factor) *
+                                double(total_size) /
+                                double(1ull << 40));
+        }
+
+        ss << "reweighting osd." << id << " to " << new_weight;
+
+        err = newcrush.adjust_item_weightf(cct, id, new_weight,
+                                           g_conf()->osd_crush_update_weight_set);
+        if (err < 0)
+          goto reply_no_propose;
+    }
+
+    if (dry_run) {
+      // return the new crush map without committing
+      p->crush->encode(rdata, mon.get_quorum_con_features());
+      ss << "test reweighted all crush weights by scaling factor "
+         << g_conf()->osd_crush_scaling_factor
+      return true;
+    }
+
+    if (!really) {
+      ss << "WARNING: this will reweight all crush weights by scaling factor "
+        << g_conf()->osd_crush_scaling_factor
+        << " and may cause data movement."
```

Comment:
How might this result in data movement, if the map update is pushed out in one atomic monolithic update?  Rounding?

Reply:
For example if someone had modified the crush weights so they are not relative to the device size.
Or if other spooky crush things happen. 
In any case -- best test it first before weighting 65 PiB of OSDs ;-)